### PR TITLE
Add a lint to suggest uint == 0 over uint <= 0

### LINKED ISSUE #592: Lint bad comparisons of unsigned ints and 0
As @scurest pointed out, we don't currently lint bad comparisons of unsigned integer types to 0. Specifically, I think we should lint `unsigned < 0`, which is always false and certainly not what the programmer intended, `unsigned <= 0` (which is weird and confusing and should be written as `== 0`) and `unsigned >= 0`, because it's a tautology and probably a bug as well.

### PR DESCRIPTION
**Pull request by scurest**
Fix #592.

### CODE REVIEW

File: src/types.rs [hunk 1]
```diff
@@ -557,3 +557,55 @@ impl LateLintPass for CharLitAsU8 {
...
 }
+
+/// **What it does:** This lint checks for expressions where an unsigned integer is tested to be non-positive and suggests testing for equality with zero instead.
+///
+/// **Why is this bad?** `x <= 0` may mislead the reader into thinking `x` can be negative. `x == 0` makes explicit that zero is the only possibility.
+///
+/// **Known problems:** None
+///
+/// **Example:** `vec.len() <= 0`
+declare_lint!(pub TEST_UINT_NONPOSITIVE, Warn,
```
Comment (Manishearth) [inline src/types.rs]: What about ABSURD_UNSIGNED_COMPARISONS?

File: src/types.rs [hunk 2]
```diff
@@ -557,3 +557,55 @@ impl LateLintPass for CharLitAsU8 {
...
+        if let ExprBinary(ref cmp, ref lhs, ref rhs) = expr.node {
+            let op = cmp.node;
+
+            let comparee = match op {
+                BiLe if is_zero_lit(rhs) => lhs, // x <= 0
+                BiGe if is_zero_lit(lhs) => rhs, // 0 >= x
+                _ => return,
+            };
+
+            if let ty::TyUint(_) = cx.tcx.expr_ty(comparee).sty {
```
Comment (Manishearth) [inline src/types.rs]: We can also look for other unsigned types, right?

File: src/types.rs [hunk 3]
```diff
@@ -557,3 +557,55 @@ impl LateLintPass for CharLitAsU8 {
...
+
+/// **What it does:** This lint checks for expressions where an unsigned integer is tested to be non-positive and suggests testing for equality with zero instead.
+///
+/// **Why is this bad?** `x <= 0` may mislead the reader into thinking `x` can be negative. `x == 0` makes explicit that zero is the only possibility.
+///
+/// **Known problems:** None
+///
+/// **Example:** `vec.len() <= 0`
+declare_lint!(pub TEST_UINT_NONPOSITIVE, Warn,
+              "tesing whether an unsigned integer is non-positive");
```
Comment (Manishearth) [inline src/types.rs]: spelling

File: src/types.rs (hunk 2 above)
Comment (fhartwig) [inline src/types.rs]: @Manishearth `TyUint` covers all the unsigned integer types. (See [here](http://manishearth.github.io/rust-internals-docs/rustc/middle/ty/enum.TypeVariants.html))

Comment (scurest) [inline src/types.rs]: So is this okay? I've added tests to make sure it works with more than one unsigned type now.

Comment (fhartwig) [inline src/types.rs]: I think this is right. `TyUint` should match all unsigned integer types, the field determines the precise type.

**Comment by Manishearth** [comment]
Thanks!

**Pull request merged**